migration: Default CK8S_DRY_RUN_INSTALL to true

simonklb opened the pull request:
<!-- Choose your PR title carefully as it will be used as the entry in the changelog! -->

<!-- markdownlint-disable MD041 -->
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [ ] personal data beyond what is necessary for interacting with this pull request, nor
> - [ ] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       <!-- This PR adds a new feature -->
- [ ] kind/improvement   <!-- This PR changes an existing feature -->
- [ ] kind/deprecation   <!-- This PR removes an existing feature -->
- [ ] kind/documentation <!-- This PR contains documentation -->
- [ ] kind/clean-up      <!-- This PR cleans up technical debt -->
- [x] kind/bug           <!-- This PR fixes a bug -->
- [ ] kind/other         <!-- This PR does something else -->

_Optional_: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change   <!-- This PR introduces an admin facing change, add "Platform Administrator notice" section -->
- [ ] kind/dev-change     <!-- This PR introduces a dev facing change, add "Application Developer notice" section -->
- [ ] kind/security       <!-- This PR introduces a critical security fix, add "Security notice" section -->
- [ ] [kind/adr](set-me\) <!-- This PR implements an ADR, add the link -->

<!-- Uncomment the additional sections that applies. -->

<!-- Additional information to be added in the release notes
### Release notes
...
-->

<!-- Additional information with kind/admin-change
### Platform Administrator notice
...
-->

<!-- Add additional information with kind/dev-change
### Application Developer notice
...
-->

<!-- Add additional information with kind/security
### Security notice
...
-->

### What does this PR do / why do we need this PR?

Without this the script fails when it's unset.

#### Information to reviewers

<!--
Any additional information reviews should know.

How to run / how to test.

Include screenshots if applicable to help explain these changes.
--->

#### Checklist

<!-- This section is not added to the changelog or release notes, it is to help you as a contributor and reviewers. -->

- [ ] Proper commit message prefix on all commits
    <!-- Example of commit message prefixes:
    - all: changes to multiple areas
    - apps: changes to applications running in all clusters
    - apps sc: changes to applications running in service clusters
    - apps wc: changes to applications running in workload clusters
    - bin: changes to management binaries
    - config: changes to configuration
    - docs: changes to documentation
    - release: release related
    - scripts: changes to scripts
    - tests: changes to tests
    --->
- Change checks:
    - [ ] The change is transparent
    - [ ] The change is disruptive
    - [ ] The change requires no migration steps
    - [ ] The change requires migration steps
    - [ ] The change updates CRDs
    - [ ] The change updates the config _and_ the schema
- Documentation checks:
    - [ ] The [public documentation](https://github.com/elastisys/welkin) required no updates
    - [ ] The [public documentation](https://github.com/elastisys/welkin) required an update - [link to change](set-me\)
- Metrics checks:
    - [ ] The metrics are still exposed and present in Grafana after the change
    - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts required no updates)
    - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts required an update)
- Logs checks:
    - [ ] The logs do not show any errors after the change
- PodSecurityPolicy checks:
    - [ ] Any changed Pod is covered by Kubernetes Pod Security Standards
    - [ ] Any changed Pod is covered by Gatekeeper Pod Security Policies
    - [ ] The change does not cause any Pods to be blocked by Pod Security Standards or Policies
- NetworkPolicy checks:
    - [ ] Any changed Pod is covered by Network Policies
    - [ ] The change does not cause any dropped packets in the NetworkPolicy Dashboard
- Audit checks:
    - [ ] The change does not cause any unnecessary Kubernetes audit events
    - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
    - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
    - [ ] The bug fix is covered by regression tests

elastisys-staffan reviewed (commented):
Nice catch! How did you find it? It is set [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/bin/upgrade.bash#L20) and [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration/template/apply/00-template.sh#L8) which I assumed would cover all cases. There are identical checks in several places, should we change them as well?

elastisys-staffan commented:
> > Nice catch! How did you find it?
> 
> While developing and running a migration snippet directly instead of via `ck8s upgrade`.
> 
> > It is set [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/bin/upgrade.bash#L20) and [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration/template/apply/00-template.sh#L8) which I assumed would cover all cases.
> 
> Any reason why you didn't add it to the prepare template? It was a prepare snippet in my case.
> 
> I've added it here: [dd51e0b](https://github.com/elastisys/compliantkubernetes-apps/commit/dd51e0ba1427d2c6bddeaf6db2c9b4d0d3827932) let me know if there is a reason for it to not be there and I'll revert it!
> 
> > There are identical checks in several places, should we change them as well?
> 
> Indeed we should! [1552d39](https://github.com/elastisys/compliantkubernetes-apps/commit/1552d39756ae4db60787f8e669a20fc1d31e9452)

The [original task](https://github.com/elastisys/compliantkubernetes-apps/issues/2733) left out the prepare step so that's why it's not there. My understanding is that the helper functions that implements dry-run should only be used for apply jobs. Prepare jobs is for local config changes using the `yq` helpers.

I'm not against adding it here as well, but I would like to hear @Ajarmar 's take on it.

simonklb commented:
> > > Nice catch! How did you find it?
> > 
> > 
> > While developing and running a migration snippet directly instead of via `ck8s upgrade`.
> > > It is set [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/bin/upgrade.bash#L20) and [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration/template/apply/00-template.sh#L8) which I assumed would cover all cases.
> > 
> > 
> > Any reason why you didn't add it to the prepare template? It was a prepare snippet in my case.
> > I've added it here: [dd51e0b](https://github.com/elastisys/compliantkubernetes-apps/commit/dd51e0ba1427d2c6bddeaf6db2c9b4d0d3827932) let me know if there is a reason for it to not be there and I'll revert it!
> > > There are identical checks in several places, should we change them as well?
> > 
> > 
> > Indeed we should! [1552d39](https://github.com/elastisys/compliantkubernetes-apps/commit/1552d39756ae4db60787f8e669a20fc1d31e9452)
> 
> The [original task](https://github.com/elastisys/compliantkubernetes-apps/issues/2733) left out the prepare step so that's why it's not there. My understanding is that the helper functions that implements dry-run should only be used for apply jobs. Prepare jobs is for local config changes using the `yq` helpers.
> 
> I'm not against adding it here as well, but I would like to hear @Ajarmar 's take on it.

cc @aarnq as well

In this specific case, I've added a "pre-flight" / acceptance test as a prepare snippet which uses (read-only) Helm commands.

simonklb commented:
Here: https://github.com/elastisys/compliantkubernetes-apps/blob/d02faf70eadf711a1d50427facc8e03588a0c075/scripts/migration/crossplane.sh#L56

And in use here: https://github.com/elastisys/compliantkubernetes-apps/blob/5c3a466f6fa364d1b987ba4e0636a658a4414d10/migration/v0.50/prepare/90-module-node-local-dns-acceptance.sh

aarnq commented:
Should the default really be to mutate things? 

Also shouldn't get / read operations be performed even if we dry-run, so why do you need to run it with false from prepare snippets given the current state of it being only implemented for things normally done in apply snippets? I think limiting it would make it safer for use in that way in prepare snippets, right? :thinking:

simonklb commented:
@elastisys-staffan @Ajarmar any last words?
